server/src/main/kotlin/org/elaastic/player/dashboard/DashboardModelFactory.kt

```diff
@@ -20,142 +20,179 @@
             sequenceId = sequence.id
         )
 
+        val learners: MutableList<LearnerMonitoringModel> = getLearnerMonitoringModels(
+            sequenceMonitoringModel,
+            sequence,
+        )
+        sequenceMonitoringModel.setLearners(learners)
+
+        val previousSequence: Sequence? = sequenceService.findPreviousSequence(sequence)
+        val nextSequence: Sequence? = sequenceService.findNextSequence(sequence)
+
+        return DashboardModel(
+            sequence,
+            previousSequence?.id,
+            nextSequence?.id,
+            sequenceMonitoringModel
+        )
+    }
+
+    private fun getLearnerMonitoringModels(
+        sequenceMonitoringModel: SequenceMonitoringModel,
+        sequence: Sequence,
+    ): MutableList<LearnerMonitoringModel> {
+        val attendees: List<LearnerAssignment> = assignmentService.getRegisteredUsers(sequence.assignment!!)
+        val responses: List<Response> = interactionService.findAllResponsesBySequenceOrderById(sequence)
+
+        // Associate each learner with the number of evaluations he made
+        val evaluationCountByUser = peerGradingService.countEvaluationsMadeByUsers(attendees, sequence)
+        val countResponseGradable = getCountResponseGradable(sequence)
+
         val learners: MutableList<LearnerMonitoringModel> = mutableListOf()
 
         attendees.forEach {
             learners.add(
-                LearnerMonitoringModel(
-                    it.learner.id!!,
-                    it.learner.getFullname(),
-                    getAttendeeStateOnResponsePhase(it, responses, sequenceMonitoringModel.phase1State),
-                    getAttendeeStateOnEvaluationPhase(
-                        it,
-                        sequence,
-                        sequenceMonitoringModel.phase2State,
-                        evaluationCountByUser,
-                        learnerToIfTheyAnswer,
-                        countResponseGradable,
-                    ),
-                    sequenceMonitoringModel = sequenceMonitoringModel
+                buildLearnerMonitoringModel(
+                    it,
+                    learnerHasAnswer(responses, it),
+                    sequenceMonitoringModel,
+                    sequence,
+                    evaluationCountByUser[it] ?: 0,
+                    countResponseGradable
                 )
             )
         }
+        return learners
+    }
 
-        sequenceMonitoringModel.setLearners(learners)
+    /**
+     * Get the number of responses that are gradable.
+     *
+     * @param sequence the sequence
+     * @return the number of responses that are gradable
+     */
+    fun getCountResponseGradable(sequence: Sequence) = try {
+        val responseStudent = responseService.findAllByAttemptNotFake(1, sequence).size.toLong()
+        val responseFake = responseService.findAllFakeResponses(sequence).size.toLong()
+        responseStudent + responseFake
+    } catch (_: IllegalStateException) {
```

Comment:
Seems to me very dangerous to catch such an exception and just consider the answer is `0`.
Is that really possible it occurs being a normal situation ? 
I would expect at least to log the exception, but I wonder why such a exception may happen here....

Reply:
In the function we call to retrieve the data, this line is called :
```kotlin
sequence.getResponseSubmissionInteraction()
```
This method from `Sequence` throw an exception when the sequence's interaction isn't loaded :
```kotlin
fun getResponseSubmissionInteraction(): Interaction {
    check(interactions.isNotEmpty()) { "No interaction defined for this sequence" }
    return interactions[InteractionType.ResponseSubmission]
        ?: throw IllegalStateException("The response submission interaction is not initialized")
}
```
As I am sure all the sequence's interactions are loaded when I get the sequence (`sequenceService.get(sequenceId, fetchInteractions  = true)`) If an exception occurs that mean the interaction really don't exist. In this case, the type of the interaction is `ResponseSubmission`. If this interaction doesn't exist, that mean no response could be written, so the number of gradable response is `0`.

Reply:
On that point, I really disagree (phew... I've started to think I was really too tired while doing the review).

The semantics here of using `IllegalStateException` is that this situation must not happen.
So you're not supposed to use a response submission interaction got through this method if it is not initialized.

I understand the behaviour may be correct, but it is semantically not correct here.

Reply:
I propose this solution :
```kotlin
fun getCountResponseGradable(sequence: Sequence): Long {
    return if (sequence.getResponseSubmissionInteractionOrNull() == null) {
        // If the sequence hasn't started yet, the ResponseSubmission interaction would be null
        // If the sequence hasn't started yet, no response has been produce,
        // so no response is gradable
        0
    } else {
        val responseStudent = responseService.findAllByAttemptNotFake(1, sequence).size.toLong()
        val responseFake = responseService.findAllFakeResponses(sequence).size.toLong()
        responseStudent + responseFake
    }
}
```
We check before trying.
What do you think ?

Reply:
Perfect